packages/fluentui/react-northstar/src/components/Tree/hooks/useTreeActiveState.tsx

```diff
@@ -0,0 +1,105 @@
+import * as React from 'react';
+import { useAutoControlled } from '@fluentui/react-bindings';
+import { TreeProps } from '../Tree';
+import { useStableProps } from './useStableProps';
+import { FlatTree } from './flattenTree';
+import { findIndex, removeItemAtIndex } from './utils';
+import * as _ from 'lodash';
+
+export interface TreeActiveState {
+  activeItemIds: string[];
+  flatTree: FlatTree;
+  toggleActive: (ids: string[], e: React.SyntheticEvent) => void;
+}
+
+export function useTreeActiveState(
+  props: Pick<TreeProps, 'defaultActiveItemIds' | 'activeItemIds' | 'exclusive'>,
+  flatTree: FlatTree,
+): TreeActiveState {
+  const initialActiveItemIds = React.useMemo(() => {
+    const initalValue = [];
+    Object.keys(flatTree).forEach(key => {
+      if (flatTree[key].expanded) {
+        initalValue.push(key);
+      }
+    });
+    return initalValue;
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, []); // initialValue only needs to be computed on mount
+
+  const [activeItemIds, setActiveItemIdsState] = useAutoControlled<string[]>({
+    defaultValue: props.defaultActiveItemIds,
+    value: props.activeItemIds,
+    initialValue: initialActiveItemIds,
+  });
+
+  // compute a new flattened tree based on activeItemIds state
+  const updatedFlatTree = React.useMemo(() => {
+    const updatedFlatTree = { ...flatTree };
+    Object.keys(updatedFlatTree).forEach(key => {
+      if (updatedFlatTree[key].expanded) {
+        updatedFlatTree[key].expanded = false;
+      }
+    });
+    activeItemIds.forEach(activeId => {
+      updatedFlatTree[activeId].expanded = true;
+    });
+    return updatedFlatTree;
+  }, [activeItemIds, flatTree]);
+
+  const toggleActiveOnOneId = React.useCallback(
+    (ids: string[], id: string): string[] => {
+      if (!updatedFlatTree[id]?.hasSubtree) {
+        // leaf node does not have the concept of active/inactive
+        return ids;
+      }
+
+      let result: string[];
+      const index = findIndex(ids, id);
+      if (index >= 0) {
+        result = removeItemAtIndex(ids, index);
```

Comment:
Why we need to use `findIndex` there?

---

- to understand if `id` exists in an array we can use `.indexOf` or `.includes`
- to remove an item from an array we can use [`_.without()`](https://lodash.com/docs/4.17.15#without)
- we don't care about ordering in `ids`

Let's remove `findIndex` and `removeItemAtIndex` if possible

Reply:
I used `findIndex` and `removeItemAtIndex` because I wasn't sure what to use, array function or lodash. So I put them in one place for easy replace.

I could change it all to `.indexOf` and `_.without()`, sounds good?
btw I found out `_.without()` is faster than slice and concat :D 
<img width="466" alt="Screenshot 2020-11-05 at 14 04 52" src="https://user-images.githubusercontent.com/28751745/98244592-f78d5e00-1f6f-11eb-8949-0da9fc6b2678.png">

Reply:
> I could change it all to .indexOf and _.without(), sounds good?

Yes, please